node/bindnode/infer.go

```diff
@@ -108,6 +112,124 @@ func inferGoType(typ schema.Type) reflect.Type {
 	panic(fmt.Sprintf("%T\n", typ))
 }
 
+// TODO(mvdan): deduplicate with inferGoType once reflect supports creating named types
+
+func produceGoType(goTypes map[string]string, typ schema.Type) (name, src string) {
+	if typ, ok := typ.(interface{ IsAnonymous() bool }); ok {
+		if typ.IsAnonymous() {
+			panic("TODO: does this ever happen?")
+		}
+	}
+
+	name = string(typ.Name())
+
+	switch typ.(type) {
+	case *schema.TypeBool:
+		return goTypeBool.String(), ""
+	case *schema.TypeInt:
+		return goTypeInt.String(), ""
+	case *schema.TypeFloat:
+		return goTypeFloat.String(), ""
+	case *schema.TypeString:
+		return goTypeString.String(), ""
+	case *schema.TypeBytes:
+		return goTypeBytes.String(), ""
+	}
+
+	// Results are cached in goTypes.
+	if src := goTypes[name]; src != "" {
+		return name, src
+	}
+
+	src = produceGoTypeInner(goTypes, name, typ)
+	goTypes[name] = src
+	return name, src
+}
+
+func produceGoTypeInner(goTypes map[string]string, name string, typ schema.Type) (src string) {
+	// Avoid infinite cycles.
+	// produceGoType will fill in the final type later.
+	goTypes[name] = "WIP"
+
+	switch typ := typ.(type) {
+	case *schema.TypeStruct:
+		var b strings.Builder
+		fmt.Fprintf(&b, "struct {\n")
+		fields := typ.Fields()
+		for _, field := range fields {
+			fmt.Fprintf(&b, "%s ", fieldNameFromSchema(field.Name()))
+			ftypGo, _ := produceGoType(goTypes, field.Type())
+			if field.IsNullable() {
+				fmt.Fprintf(&b, "*")
+			}
+			if field.IsOptional() {
+				fmt.Fprintf(&b, "*")
+			}
+			fmt.Fprintf(&b, "%s\n", ftypGo)
+		}
+		fmt.Fprintf(&b, "\n}")
+		return b.String()
+	case *schema.TypeMap:
+		ktyp, _ := produceGoType(goTypes, typ.KeyType())
+		vtyp, _ := produceGoType(goTypes, typ.ValueType())
+		if typ.ValueIsNullable() {
+			vtyp = "*" + vtyp
+		}
+		return fmt.Sprintf(`struct {
+			Keys []%s
+			Values map[%s]%s
+		}`, ktyp, ktyp, vtyp)
+	case *schema.TypeList:
+		etyp, _ := produceGoType(goTypes, typ.ValueType())
+		if typ.ValueIsNullable() {
+			etyp = "*" + etyp
+		}
+		return fmt.Sprintf("[]%s", etyp)
+	case *schema.TypeUnion:
+		var b strings.Builder
+		fmt.Fprintf(&b, "struct{\n")
+		members := typ.Members()
+		for _, ftyp := range members {
+			ftypGo, _ := produceGoType(goTypes, ftyp)
+			fmt.Fprintf(&b, "%s ", fieldNameFromSchema(string(ftyp.Name())))
+			fmt.Fprintf(&b, "*%s\n", ftypGo)
+		}
+		fmt.Fprintf(&b, "\n}")
+		return b.String()
+	}
+	panic(fmt.Sprintf("%T\n", typ))
+}
+
+func ProduceGoTypes(w io.Writer, schemaTypes map[schema.TypeName]schema.Type) error {
```

Comment:
oh, WOW!  Minicodegen!  Sweet!  I had been thinking this would be maybe kinda neat to have in our back pockets, and now here it is!

I wonder if it should be shifted to a subpackage?  Obviously it's not mechanically necessary to do so, but also it seems like a sufficiently different kind of feature than all the other publicly exported functions of this package that it might be worth it for organization and messaging.

Reply:
I wrote it in a couple of hours because I'm too lazy to spend the two hours instead writing the types by hand :)

I was also thinking about where to put this. Ideally it would be a program, like `node/bindnode/bindnodegen` or something like that. I don't think that's particularly useful until we have a "parse schema file" library API, though. For now, I think we can just put this behind a `// +build bindnodegen` to not expose it in the bindnode API, but still allow accessing it form other packages like:

```
//go:generate go test -run TestGenerate -tags bindnodegen
```

Reply:
> I wrote it in a couple of hours because I'm too lazy to spend the two hours instead writing the types by hand :)

:raised_hands: 
Sounds like it was extremely worth it.

I'm super happy about this; it'll make this immediately less of a pain to maintain if we tweak the schema-schema.

> node/bindnode/bindnodegen

I advocate you do this immediately :)

Even if we don't wrap a full CLI around it today, I can already think of another project where I'd very likely use this even in its nascent form.